### PR TITLE
Connectivity components in ST

### PR DESCRIPTION
**Pull request by Dantara**
I implemented an algorithm that we discussed (_finding connectivity components by using `union-find`_).
The implementation doesn't work properly but it seems like exactly what we discussed. 
I will try to fix it soon.

### CODE REVIEW

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 1]
```diff
@@ -1,18 +1,73 @@
...
+pointsList :: [(a, Point s a)] -> ST s [(a, [a])]
+pointsList = mapM $ \(x, px) -> do
+  repr <- UF.descriptor px
+  return (repr, [x])
+
+-- | O(s).
+mkVertexPoints :: Eq a => Graph a -> ST s ([(a, Point s a)], (Graph (Point s a)))
+mkVertexPoints Empty = return ([], Empty)
+mkVertexPoints (Vertex v) = do
+  point <- UF.fresh v
```
Comment (fizruk) [inline src/Algebra/Graph/Algorithm/Internal.hs:40]: You need to check if you already created a point for this vertex. Something like this:

```suggestion
  case Map.lookup v points of
    Nothing -> do
      point <- UF.fresh v
      modify (Map.insert v point)
      return (Vertex Point)
```

So it makes sense to do it as `State (Map a (Point s a)) (Graph (Point s a))` and then to convert `Map` back into a list of pairs.

For `Ord a`, `Map` is good, but perhaps for integers from a relatively small range it is worth making a version that uses `Vector (Point s a)` instead.

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 2]
```diff
@@ -33,19 +34,22 @@ pointsList = mapM $ \(x, px) -> do
...
+mkVertexPoints (Connect l r) ps = do
+  (lp, lg) <- mkVertexPoints l ps
+  (rp, rg) <- mkVertexPoints r ps
   return (lp <> rp, Connect lg rg)
-mkVertexPoints (Overlay l r) = do
-  (lp, lg) <- mkVertexPoints l
-  (rp, rg) <- mkVertexPoints r
+mkVertexPoints (Overlay l r) ps = do
+  (lp, lg) <- mkVertexPoints l ps
+  (rp, rg) <- mkVertexPoints r ps
```
Comment (Dantara) [inline src/Algebra/Graph/Algorithm/Internal.hs]: I replaced `[(a, Point s a]` with `Map a (Point s a)`. I tried to do it inside the `State` monad but got compilation error. After it rewrote with explicit state passing. I will try to fix the compilation error soon.

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 3]
```diff
@@ -14,12 +15,12 @@ import qualified Data.UnionFind.ST as UF
 -- **NOTE:** Doesn't work properly. Check an example.
 --
 -- >>> components (((1 * 2) + (3 * 4)) + (2 * 3))
--- fromList [(2,[2,1]),(3,[3,2]),(4,[4,3])]
+-- fromList [(2,[2,1]),(4,[4,3])]
```
Comment (Dantara) [inline src/Algebra/Graph/Algorithm/Internal.hs:18]: In this version finial answer is different but still incorrect.

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 4]
```diff
@@ -1,18 +1,77 @@
...
+      return (Map.insert v p ps, Vertex p)
+    _ -> return (ps, Empty)
+mkVertexPoints (Connect l r) ps = do
+  (lp, lg) <- mkVertexPoints l ps
+  (rp, rg) <- mkVertexPoints r ps
+  return (lp <> rp, Connect lg rg)
+mkVertexPoints (Overlay l r) ps = do
+  (lp, lg) <- mkVertexPoints l ps
+  (rp, rg) <- mkVertexPoints r ps
+  return (lp <> rp, Overlay lg rg)
```
Comment (fizruk) [inline src/Algebra/Graph/Algorithm/Internal.hs:53]: State should be threaded. When you consider right subgraph, you should account for points that have been created on the left:

```suggestion
  (lp, lg) <- mkVertexPoints l ps
  (rp, rg) <- mkVertexPoints r lp
  return (rp, Overlay lg rg)
```

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 5]
```diff
@@ -1,18 +1,77 @@
...
+mkVertexPoints (Vertex v) ps =
+  case Map.lookup v ps of
+    Nothing -> do
+      p <- UF.fresh v
+      return (Map.insert v p ps, Vertex p)
+    _ -> return (ps, Empty)
+mkVertexPoints (Connect l r) ps = do
+  (lp, lg) <- mkVertexPoints l ps
+  (rp, rg) <- mkVertexPoints r ps
+  return (lp <> rp, Connect lg rg)
```
Comment (fizruk) [inline src/Algebra/Graph/Algorithm/Internal.hs:49]: Same as below:

```suggestion
  (lp, lg) <- mkVertexPoints l ps
  (rp, rg) <- mkVertexPoints r lp
  return (rp, Connect lg rg)
```

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 6]
```diff
@@ -45,12 +45,12 @@ mkVertexPoints (Vertex v) ps =
     _ -> return (ps, Empty)
 mkVertexPoints (Connect l r) ps = do
   (lp, lg) <- mkVertexPoints l ps
-  (rp, rg) <- mkVertexPoints r ps
-  return (lp <> rp, Connect lg rg)
+  (rp, rg) <- mkVertexPoints r lp
+  return (rp, Connect lg rg)
```
Comment (Dantara) [inline src/Algebra/Graph/Algorithm/Internal.hs:49]: Fixed problem in mkVertexPoints but error still remaining.

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 7]
```diff
@@ -1,18 +1,77 @@
...
+
+-- | O(s + n log n).
+mkVertexPoints :: Ord a => Graph a -> Map a (Point s a) -> ST s (Map a (Point s a), Graph (Point s a))
+mkVertexPoints Empty _ = return (Map.empty, Empty)
+mkVertexPoints (Vertex v) ps =
+  case Map.lookup v ps of
+    Nothing -> do
+      p <- UF.fresh v
+      return (Map.insert v p ps, Vertex p)
+    _ -> return (ps, Empty)
```
Comment (fizruk) [inline src/Algebra/Graph/Algorithm/Internal.hs:45]: Why are you changing the graph? 🙂

File: src/Algebra/Graph/Algorithm/Internal.hs [hunk 8]
```diff
@@ -1,18 +1,77 @@
...
+componentsST g@(Connect _ _) = do
+  points <- extractPointsST g
+  case points of
+    (p:ps) ->
+      foldM_ (\x y -> (UF.union x y) >> (return y)) p ps
+    _ -> return ()
+componentsST _ = return ()
+
+-- | O(s).
+extractPointsST :: Graph (Point s a) -> ST s [Point s a]
```
Comment (unaimillan) [inline src/Algebra/Graph/Algorithm/Internal.hs:67]: Is it possible to get rid of `ST s` context here and have signature like `Graph (Point s a) -> [Point s a] `?

File: src/Algebra/Graph/Algorithm/Internal.hs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Dantara) [reply src/Algebra/Graph/Algorithm/Internal.hs]: Fixed this but answer still incorrect.

File: src/Algebra/Graph/Algorithm/Internal.hs (hunk 8 above)
Comment (Dantara) [reply src/Algebra/Graph/Algorithm/Internal.hs:67]: I guess it is not possible. I think `s` in `Point s a` should always be connected with `s` in `ST s` to track context.

**Pull request merged**